chain/client/src/stateless_validation/partial_witness_tracker.rs

```diff
@@ -0,0 +1,177 @@
+                "Received duplicate or redundant partial state witness part."
+            );
+            return None;
+        }
+
+        // Check if we have already decoded the state witness.
+        if self.is_decoded {
+            return None;
+        }
+
+        // Insert the part into the cache entry.
+        self.parts[part_ord] = Some(part);
+        self.data_parts_present += 1;
+
+        // If we have enough parts, try to decode the state witness.
+        if self.data_parts_present < self.data_parts_required {
+            return None;
+        }
+        self.is_decoded = true;
+        match rs.decode(&mut self.parts, encoded_length) {
+            Ok(encoded_chunk_state_witness) => Some(encoded_chunk_state_witness),
+            Err(err) => {
+                // We ideally never expect the decoding to fail. In case it does, we received a bad part
+                // from the chunk producer.
+                tracing::error!(target: "stateless_validation", ?err, ?chunk_hash, "Failed to decode witness part.");
+                None
+            }
+        }
+    }
+}
+
+/// Track the Reed Solomon erasure encoded parts of the `EncodedChunkStateWitness`. These are created
+/// by the chunk producer and distributed to validators. Note that we do not need all the parts of to
+/// recreate the full state witness.
+pub struct PartialEncodedStateWitnessTracker {
+    /// Epoch manager to get the set of chunk validators
+    epoch_manager: Arc<dyn EpochManagerAdapter>,
+    /// Keeps track of state witness parts received from chunk producers.
+    parts_cache: LruCache<ChunkHash, CacheEntry>,
+    /// Reed Solomon encoder for decoding state witness parts.
+    rs_map: RsMap,
+}
+
+impl PartialEncodedStateWitnessTracker {
+    pub fn new(epoch_manager: Arc<dyn EpochManagerAdapter>) -> Self {
+        Self {
+            epoch_manager,
+            parts_cache: LruCache::new(NUM_CHUNKS_IN_WITNESS_TRACKER_CACHE),
+            rs_map: RsMap::new(),
+        }
+    }
+
+    pub fn store_partial_encoded_state_witness(
+        &mut self,
+        partial_witness: PartialEncodedStateWitness,
+    ) -> Result<(), Error> {
+        let chunk_validators = self.epoch_manager.get_chunk_validator_assignments(
+            &partial_witness.epoch_id(),
+            partial_witness.chunk_header().shard_id(),
+            partial_witness.chunk_header().height_created(),
+        )?;
+        let rs = self.rs_map.entry(chunk_validators.len());
+
+        PartialEncodedStateWitnessTracker::maybe_insert_new_entry_in_parts_cache(
+            &partial_witness,
+            &mut self.parts_cache,
+            &rs,
+        );
+
+        let chunk_hash = partial_witness.chunk_header().chunk_hash();
+        let entry = self.parts_cache.get_mut(&chunk_hash).unwrap();
+
+        if let Some(_witness) = entry.insert_in_cache_entry(partial_witness, rs) {
+            // TODO(stateless_validation): Send encoded state witness to client here.
```

Comment:
This will be part of follow up PR. We need to pipe through a state witness actor to client actor message...